### PR TITLE
python3Packages.datatable: fix darwin build

### LINKED ISSUE #65611: pythonPackages.datatable is failing to build on Darwin
**Describe the bug**

Unable to build python3Packages.datatable.

**To Reproduce**
1. `nix-build -A python3Packages.datatable`

**Expected behavior**
Derivation builds correctly

**Additional context**

<details>
<summary>Build error log</summary>

```
 λ nix-build -A python3Packages.datatable
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
these derivations will be built:
  /nix/store/nj0fakfpf6adxzsq7yx2lhf64afdrv4s-python3.7-datatable-0.8.0.drv
building '/nix/store/nj0fakfpf6adxzsq7yx2lhf64afdrv4s-python3.7-datatable-0.8.0.drv'...
unpacking sources
unpacking source archive /nix/store/6sv9c1j377k9gb2s62zsg7ph2w4hxhkz-datatable-0.8.0.tar.gz
source root is datatable-0.8.0
setting SOURCE_DATE_EPOCH to timestamp 1549398170 of file datatable-0.8.0/setup.cfg
patching sources
configuring
building

Start setup.py
  command = `bdist_wheel`

Determine the compiler
  Using compiler from environment variable `CXX`: clang++

Prepare the environment
  CC = clang++
  CXX = clang++
  LDFLAGS =
  ARCHFLAGS = -m64
  LLVM_CONFIG =
  MACOSX_DEPLOYMENT_TARGET = 10.10

Find an LLVM installation
  Environment variable LLVM = /nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0
  Llvm directory: /nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0
  Version: llvm7

Find compile include directories
  `c` is the main C++ source directory
  `/nix/store/7z8qn9wsilmsa92ipc1qvg5zq9npi9lm-python3-3.7.4/include/python3.7m` added from CONFINCLUDEPY
  `/nix/store/7z8qn9wsilmsa92ipc1qvg5zq9npi9lm-python3-3.7.4/include` added from sys.prefix
  `/nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0/include` added from Llvm package
  `/nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0/include/c++/v1` added from Llvm package
  Directory `/nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0/include/c++/v1` not found, ignoring

Determine the extra compiler flags
  -fopenmp
  -std=c++11
  -stdlib=libc++
  -Ic
  -fPIC
  -O3
  -g0
  -Weverything
  -Wno-c++98-compat-pedantic
  -Wno-c99-extensions
  -Wno-exit-time-destructors
  -Wno-float-equal
  -Wno-global-constructors
  -Wno-switch-enum
  -Wno-weak-template-vtables
  -I/nix/store/7z8qn9wsilmsa92ipc1qvg5zq9npi9lm-python3-3.7.4/include
  -isystem /nix/store/7z8qn9wsilmsa92ipc1qvg5zq9npi9lm-python3-3.7.4/include
  -I/nix/store/7z8qn9wsilmsa92ipc1qvg5zq9npi9lm-python3-3.7.4/include/python3.7m
  -isystem /nix/store/7z8qn9wsilmsa92ipc1qvg5zq9npi9lm-python3-3.7.4/include/python3.7m
  -I/nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0/include
  -isystem /nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0/include
  -Ic

Find the required dynamic libraries
  /nix/store/m9zwymfk81y2vvwjbsb4hn2pjcq8s22m-llvm-7.1.0/lib/libomp.dylib does not exist

Determine the extra linker flags


SystemExit: `locate` command returned the following error:


builder for '/nix/store/nj0fakfpf6adxzsq7yx2lhf64afdrv4s-python3.7-datatable-0.8.0.drv' failed with exit code 1
```

</details>

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.

 - system: `"x86_64-darwin"`
 - host os: `Darwin 18.6.0, macOS 10.14.5`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.2.2`
 - channels(yl): `"darwin, nixpkgs-19.09pre185256.5b8b5a694c7"`
 - nixpkgs: `/run/current-system/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: pythonPackages.datatable
# a list of nixos modules affected by the problem
module:
```

cc @abbradar @jonringer @FRidh

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
Attempt to resolve #65611

Need to be at home before I can verify on my own mac.

Verified it doesn't affect linux derivation.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @kalbasit

### CODE REVIEW

**Comment by jonringer** [comment]
@GrahamcOfBorg build python3Packages.datatable

**Comment by jonringer** [comment]
keep forgetting that ofBorg doesn't build darwin anymore, darn :(

**Comment by kalbasit** [comment]
I'm still getting the same error.

**Comment by worldofpeace** [comment]
> keep forgetting that ofBorg doesn't build darwin anymore, darn :(

I believe that's because you need to be a trusted user to trigger darwin builds.
It may be a good idea to submit that at the ofborg repo if you plan on regularly
contributing to the darwin infrastructure in nixpkgs.

**Comment by jonringer** [comment]
@worldofpeace I don't trust me. :)

**Comment by kalbasit** [comment]
@GrahamcOfBorg build python3Packages.datatable

File: pkgs/development/python-modules/datatable/hardcode-library-paths.patch [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+-                    assert os.path.isfile(fullpath), "Invalid path: %r" % (fullpath,)
+-                    resolved.append(fullpath)
+-                    log.info("Library `%s` found at %s" % (libname, fullpath))
+-                    continue
+-                else:
+-                    log.fatal("Cannot locate dynamic library `%s`" % libname)
+-            else:
+-                log.fatal("`locate` command returned the following error:\n%s"
+-                          % stderr.decode())
++        resolved = ["@libomp.dylib@", "@libc++.1.dylib@", "@libc++abi.dylib@"]
```
Comment (worldofpeace) [inline pkgs/development/python-modules/datatable/hardcode-library-paths.patch]: This is dangerous if they're not super valid bash variables.

Probably better to be `libomp_dylib` etc.

File: pkgs/development/python-modules/datatable/default.nix [hunk 2]
```diff
@@ -17,10 +21,27 @@ buildPythonPackage rec {
...
   };
 
+  patches = [
+    # Disable the compiler monkey patching, and remove the task that's copying
+    # the native dependencies to the build directory.
+    ./remove-compiler-monkeypatch_disable-native-relocation.patch
+  ] ++ lib.optionals stdenv.isDarwin [
+    # Replace the library auto-detection with hardcoded paths. These paths are
+    # then replaced by the following postPatch.
+    ./hardcode-library-paths.patch
```
Comment (worldofpeace) [inline pkgs/development/python-modules/datatable/default.nix]: You can actually do
```
(substituteAll {
  src = ./hardcode-library-paths.patch;
  libomp_dylib = "${lib.getLib openmp}/lib/libomp.dylib";
})
```

in patches which I feel works much cleaner

**Review by worldofpeace — CHANGES_REQUESTED** [review]

**Comment by kalbasit** [comment]
@worldofpeace PTAL

@GrahamcOfBorg build python3Packages.datatable

**Comment by jonringer** [comment]
We may be able to edit/remove the buildInputs addition. Nix should be able to detect that there are store paths. Are you able to run this package with a normal workload?

**Comment by kalbasit** [comment]
> We may be able to edit/remove the buildInputs addition. Nix should be able to detect that there are store paths. Are you able to run this package with a normal workload?

I have not tried doing that. It was failing for me due to the lack of `shutil` and I noticed that it was attempting to copy native binaries to the build directory which useless anyway.

**Comment by jonringer** [comment]
Removing openmp breaks the build, but I may the add the others since I'm assuming they will be needed at runtime anyway.

Other than that, I don't have a good way to test this package actually works :(

**Comment by kalbasit** [comment]
> Removing openmp breaks the build, but I may the add the others since I'm assuming they will be needed at runtime anyway.
> 
> Other than that, I don't have a good way to test this package actually works :(

The tests pass. I've also checked the linked dynamic libraries and they look good.

**Pull request merged**